src/main/java/org/opentest4j/fixtures/FixtureManager.java

```diff
@@ -0,0 +1,256 @@
+		}
+	}
+
+	/**
+	 * Resets this fixture to its initial state.
+	 */
+	public void reset() {
+		state = State.HEALTHY;
+		tearDowns.clear();
+		testPostconditions.clear();
+	}
+
+	/**
+	 * Adds a {@link TearDown} to run after the test completes.
+	 */
+	public void addTearDown(TearDown tearDown) {
+		requireNonNull(tearDown);
+		state.checkCanModifyFixture();
+		tearDowns.add(tearDown);
+	}
+
+	/**
+	 * Runs all of the {@link TearDown} instances managed by this class.
+	 *
+	 * @param errors
+	 *            filled in with a list of all errors encountered while running
+	 *            the tear downs.
+	 */
+	public void runAllTearDowns(List<Throwable> errors) {
+		requireNonNull(errors);
+		state = state.startRunTearDowns();
+
+		while (!tearDowns.isEmpty()) {
+			try {
+				tearDowns.pop().tearDown();
+			}
+			catch (Throwable e) {
+				errors.add(e);
+			}
+		}
+		state = State.COMPLETED;
+	}
+
+	/**
+	 * Runs the {@link TestPostcondition} instances managed by this class. If
+	 * any of the postconditions throws an exception, this method will throw
+	 * that exception. If no postconditions throw an exception, then all
+	 * postconditions will be callled.
+	 *
+	 * @throws Exception
+	 *             exception thrown from first failing postcondition
+	 */
+	public void runAllPostconditions() throws Exception {
+		state = state.startRunPostConditions();
+
+		try {
+			for (TestPostcondition postcondition : testPostconditions) {
+				postcondition.verify();
+			}
+		}
+		finally {
+			testPostconditions.clear();
+		}
+	}
+
+	private FixtureManager(ClassWrapper testClass, InstanceMethod testMethod) {
+		this.testMethod = testMethod;
+		this.testClass = testClass;
+	}
+
+	private class FixtureContextFacade extends FixtureContext {
+		private boolean initialized = false;
+
+		public void initialize(TestFixture testFixture) throws Throwable {
+			if (initialized) {
+				throw new IllegalStateException("Already initialized");
+			}
+			try {
+				testFixture.initialize(this);
+			}
+			finally {
+				initialized = true;
+			}
+		}
+
+		@Override
+		public void addTearDown(TearDown tearDown) {
+			requireNonNull(tearDown);
+			state.checkCanModifyFixture();
+			tearDowns.add(tearDown);
+		}
+
+		@Override
+		public void addTestPostcondition(TestPostcondition postcondition) {
+			requireNonNull(postcondition);
+			state.checkCanModifyFixture();
+			testPostconditions.add(postcondition);
+		}
+
+		@Override
+		public InstanceMethod getInstanceMethod() {
+			return testMethod;
+		}
+
+		@Override
+		public ClassWrapper getTestClass() {
+			return testClass;
+		}
+	}
+
+	private enum State {
+		HEALTHY {
+			@Override
+			void checkCanModifyFixture() {
+			}
+
+			@Override
+			State startRunPostConditions() {
+				return READY_FOR_TEARDOWN;
+			}
+
+			@Override
+			State startRunTearDowns() {
+				return RUNNING_TEAR_DOWNS;
+			}
+		},
+		READY_FOR_TEARDOWN {
+			@Override
+			void checkCanModifyFixture() {
+				throw new IllegalStateException();
```

Comment:
Just throwing `IllegalStateException` without message is not very helpful for the user. (same goes for all other instances in this file)